lib/protocol/encryption.go

```diff
@@ -0,0 +1,521 @@
+	minPaddedSize         = 1024 // smallest block we'll allow
+	blockOverhead         = tagSize + nonceSize
+	maxPathComponent      = 200              // characters
+	encryptedDirExtension = ".syncthing-enc" // for top level dirs
+	miscreantAlgo         = "AES-SIV"
+)
+
+// The encryptedModel sits between the encrypted device and the model. It
+// receives encrypted metadata and requests from the untrusted device, so it
+// must decrypt those and answer requests by encrypting the data.
+type encryptedModel struct {
+	Model
+	folderKeys map[string]*[keySize]byte // folder ID -> key
+}
+
+func (e encryptedModel) Index(deviceID DeviceID, folder string, files []FileInfo) error {
+	if folderKey, ok := e.folderKeys[folder]; ok {
+		// incoming index data to be decrypted
+		if err := decryptFileInfos(files, folderKey); err != nil {
+			return err
+		}
+	}
+	return e.Model.Index(deviceID, folder, files)
+}
+
+func (e encryptedModel) IndexUpdate(deviceID DeviceID, folder string, files []FileInfo) error {
+	if folderKey, ok := e.folderKeys[folder]; ok {
+		// incoming index data to be decrypted
+		if err := decryptFileInfos(files, folderKey); err != nil {
+			return err
+		}
+	}
+	return e.Model.IndexUpdate(deviceID, folder, files)
+}
+
+func (e encryptedModel) Request(deviceID DeviceID, folder, name string, blockNo, size int32, offset int64, hash []byte, weakHash uint32, fromTemporary bool) (RequestResponse, error) {
+	folderKey, ok := e.folderKeys[folder]
+	if !ok {
+		return e.Model.Request(deviceID, folder, name, blockNo, size, offset, hash, weakHash, fromTemporary)
+	}
+
+	// Figure out the real file name, offset and size from the encrypted /
+	// tweaked values.
+
+	realName, err := decryptName(name, folderKey)
+	if err != nil {
+		return nil, err
+	}
+	realSize := size - blockOverhead
+	realOffset := offset - int64(blockNo*blockOverhead)
+
+	if size < minPaddedSize {
+		return nil, errors.New("short request")
+	}
+
+	// Perform that request and grab the data. Explicitly zero out the
+	// hashes which are meaningless.
+
+	resp, err := e.Model.Request(deviceID, folder, realName, blockNo, realSize, realOffset, nil, 0, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encrypt the response. Blocks smaller than minPaddedSize are padded
+	// with random data.
+
+	data := resp.Data()
+	if len(data) < minPaddedSize {
+		nd := make([]byte, minPaddedSize)
+		copy(nd, data)
+		if _, err := rand.Read(nd[len(data):]); err != nil {
+			panic("catastrophic randomness failure")
+		}
+		data = nd
+	}
+	fileKey := FileKey(realName, folderKey)
+	enc := encryptBytes(data, fileKey)
+	resp.Close()
+	return rawResponse{enc}, nil
+}
+
+func (e encryptedModel) DownloadProgress(deviceID DeviceID, folder string, updates []FileDownloadProgressUpdate) error {
+	if _, ok := e.folderKeys[folder]; !ok {
+		return e.Model.DownloadProgress(deviceID, folder, updates)
+	}
+
+	// The updates contain nonsense names and sizes, so we ignore them.
+	return nil
```

Comment:
I guess we can decrypt the names and adjust the sizes?

Or otherwise not send them at all?

Reply:
This originates from when the other side folder was just a normal one getting odd data, at this point it's a separate folder type so we should indeed not even send them.